spec/graphlient/client_schema_spec.rb

```diff
@@ -1,18 +1,79 @@
 require 'spec_helper'
+require 'tempfile'
 
 describe Graphlient::Client do
-  let(:client) { Graphlient::Client.new('http://graph.biz/graphql') }
+  let(:client) { described_class.new('http://graph.biz/graphql') }
+
+  let(:url) { 'http://graph.biz/graphql' }
 
   describe '#schema' do
-    before do
-      stub_request(:post, 'http://graph.biz/graphql').to_return(status: 500)
+    context 'when server returns error' do
+      before do
+        stub_request(:post, 'http://graph.biz/graphql').to_return(status: 500)
+      end
+
+      it 'fails with an exception' do
+        expect do
+          client.schema
+        end.to raise_error Graphlient::Errors::ServerError do |e|
+          expect(e.to_s).to eq 'the server responded with status 500'
+        end
+      end
+    end
+
+    context 'when introspection request is sucessfull' do
+      before do
+        stub_request(:post, url)
+          .to_return(body: DummySchema.execute(GraphQL::Introspection::INTROSPECTION_QUERY).to_json)
+      end
+
+      it 'returns Grapql::Schema instance' do
+        expect(client.schema).to be_a(GraphQL::Schema)
+      end
+    end
+  end
+
+  describe '#dump_schema' do
+    context 'when schema path is not given' do
+      it 'raises error' do
+        expect { client.dump_schema }.to raise_error(Graphlient::Client::MissingConfigurationError)
+      end
     end
 
-    it 'fails with an exception' do
-      expect do
-        client.schema
-      end.to raise_error Graphlient::Errors::ServerError do |e|
-        expect(e.to_s).to eq 'the server responded with status 500'
+    context 'when schema path is given' do
+      let(:client) { described_class.new('http://graph.biz/graphql', schema_path: @schema_path) }
+
+      let!(:introspection_query_request) do
+        stub_request(:post, url)
+          .with(body: /query IntrospectionQuery/)
+          .to_return(body: DummySchema.execute(GraphQL::Introspection::INTROSPECTION_QUERY).to_json)
+      end
+
+      let(:schema_path) { @schema_path }
+
+      around(:each) do |example|
+        Tempfile.open('graphql_schema.json') do |file|
+          @schema_path = file.path
+          example.run
+          @schema_path = nil
+        end
+      end
+
+      it 'makes introspection query' do
+        client.dump_schema
+        expect(introspection_query_request).to have_been_made.once
+      end
+
+      it 'updates schema json file' do
+        expect { client.dump_schema }.to(change { File.read(@schema_path) })
+      end
+
+      context 'when path is instance of Pathname' do
+        let(:schema_path) { Pathname.new(@schema_path) }
+
+        it 'does not raise error' do
+          expect { client.dump_schema }.not_to raise_error
+        end
```

Comment:
This entire spec looks a bit overcomplicated... Any chance you could simplify this by using less `context`s?